### ISSUE TITLE
[Install Failure in Macbook Pro] nvcc fatal : redefinition of argument 'std'

### ISSUE THREAD

**Issue by ScartleRoy** [dmlc/mxnet#3363]
Hi,

I was trying to install MXNet in my Macbook Pro (OS X 10.11.6, GT 750M) with GPU support, but the same error occurs in several trials:

```
 >build/src/ndarray/ndarray_function_gpu.d
nvcc fatal   : redefinition of argument 'std'
make: *** [build/src/ndarray/ndarray_function_gpu.o] Error 1
```

My osx.mk file is:

```
#---------------------
# choice of compiler
#--------------------

export CC = clang-omp
export CXX = clang-omp++
export NVCC = nvcc

# whether compile with debug
DEBUG = 0

# the additional link flags you want to add
# ADD_LDFLAGS = -L/usr/local/opt/openblas/lib -L/usr/local/cuda/lib -L/usr/local/lib

# the additional compile flags you want to add
# ADD_CFLAGS = -I/usr/local/opt/openblas/include -I/usr/local/include

#---------------------------------------------
# matrix computation libraries for CPU/GPU
#---------------------------------------------

# whether use CUDA during compile
USE_CUDA = 1

# add the path to CUDA library to link and compile flag
# if you have already add them to environment variable, leave it as NONE
# USE_CUDA_PATH = /usr/local/cuda
USE_CUDA_PATH = /Developer/NVIDIA/CUDA-7.5

# whether use CUDNN R3 library
USE_CUDNN = 1

# whether use cuda runtime compiling for writing kernels in native language (i.e. Python)
USE_NVRTC = 1

# whether use opencv during compilation
# you can disable it, however, you will not able to use
# imbin iterator
USE_OPENCV = 1

# use openmp for parallelization
USE_OPENMP = 0

# choose the version of blas you want to use
# can be: mkl, blas, atlas, openblas
USE_BLAS = apple
```

I guess there might be something wrong within CMakeList.txt, but I'm not familiar with it so I don't know the exact reason. Could you please take a look into this?

BTW, this problem occurs no matter how I set ```USE_CUDA```, ```USE_CUDNN``` and ```USE_NVRTC```.

**Comment by piiswrong** [comment]
cmake is for windows.
follow this to build on ios http://mxnet.dmlc.ml/how_to/build.html#building-on-osx

**Comment by piiswrong** [comment]
@yajiedesign Can we add a warning in cmakelist to make it fail on linux and osx?

**Comment by ScartleRoy** [comment]
I followed that instruction and also failed multiple times. I searched for several issues, and still didn't work.

**Comment by ScartleRoy** [comment]
Oh one thing to correct: I fixed ```USE_CUDA = 1``` for installing a GPU-enabled version. If I set it to 0 (and ```USE_CUDNN = 0, USE_NVRTC = 0```), it came up with:
```
ld: symbol(s) not found for architecture x86_64
clang-3.5: error: linker command failed with exit code 1 (use -v to see invocation)
```
which I have seen in other issues. Spent hours on it and still haven't successfully installed...

**Comment by yajiedesign** [comment]
@piiswrong yes
` If(NOT WIN32)` 
` message( FATAL_ERROR "You can not build in this os, CMake will exit." )` 
` endif()`

**Comment by ScartleRoy** [comment]
Well, after tons of trails I finally make that work. Here are something I found useful:

1. Use homebrew to upgrade everything installed (sometimes needs to fix the links).

2. Use LLVM-3.7 and add the corresponding flags. v3.8 and v3.5 failed in my laptop.

3. Set ```USE_CUDA_PATH = /usr/local/cuda``` instead of ```USE_CUDA_PATH = /Developer/NVIDIA/CUDA-7.5```. I don't know why the latter one leaded to errors.

4. Set ```USE_BLAS = apple```. I'm not sure if openblas would work - I haven't tried that.

These configurations solved multiple errors I encountered, e.g., ```ld: symbol(s) not found for architecture x86_64```, ```nvcc fatal   : redefinition of argument 'std'```, issues about ```im2rec```, ```llvm identifier "_Allocator" is undefined```,  and many others. 

After that, I set ```USE_CUDA = 1, USE_CUDNN = 1, USE_NVRTC = 0, USE_OPENCV = 1, USE_OPENMP = 0```, and the installation succeeded. I could run the examples in both CPU and GPU mode.

Hope that this could help someone, since it's really annoying. I would appreciate if the documents and installation guide could be more clear.
